### ISSUE TITLE
Katas

### ISSUE THREAD

**Issue by OleMchls** [hh-ex/planning#15]
## Idea

A kata is a self-contained coding problem in one GitHub repository. In the past, we have used katas as a great group non-talk activity. The feedback always have been very positive, so we would like to do more.

## How to prepare a kata
A kata is a self-contained coding problem in one GitHub repository. Preferably in the `hh-ex` namespace (ask organizers for permissions, you can also move repos laters, start right away!). 

The ideal meetup coding kata problem...
- has little to no business logic, so it can be understood with little context as possible.
- can be solved by all skill levels within 60 to 90 minutes
- can be solved in iterations rather than a big a _big bang release_
- has no external network dependencies
- can easily be tested

Each kata should contain a `readme` describing the problem including some examples. It also has some general code setup, like a running elixir project, incl `Mixfile`, and the usual folder structure, so that attendees can start implementing code right away.

Ideally, a kata also provides failing specs. If it has specs, set up PR tests any CI tool of your choice, so when we are comparing results we will see if a solution passes the tests.

And if you are lacking a good kata idea, take a look here: http://rosettacode.org/wiki/Category:Programming_Tasks 

## How to run a Kata
In addition to the usual rules on how to make a meetup great for everyone, make it open and inclusive, etc. Katas works best when we pair up the group into teams. These teams are formed best when assigning team members at random (shuffle).

A nice way to shuffle group members was introduced to us by @pierrebeitz, by ripping apart a zine page and make attendees find their corresponding mates by finding the same page parts.

## Past Katas
- [List Span](https://github.com/hh-ex/kata-list-span)
- [List Ops](https://github.com/hh-ex/kata-list-ops)
- [Sleep Sort](https://github.com/hh-ex/kata-sleepsort)
- [dtgcaa](https://github.com/hh-ex/kata-dtgcaa)
- [binary matching](https://github.com/hh-ex/kata-binary)
- binary rfc sigils _tbd_
- million coin miner _tbd_

## Topic Wishlist
- General Elixir concepts (_special_ data structs, processes, GenServers)

**Comment by KlausTrainer** [comment]
Hi everybody! I've created a new kata, and I'm looking for feedback: https://github.com/hh-ex/kata-dtgcaa

Is it too short, too easy, or too hard? Does it require too much context?

I'd be happy about your feedback!

**Comment by h4cc** [comment]
@KlausTrainer Very interesting and usefull Kata, maybe some more inline docs might help understanding private functions faster. Took me some time reading the code to get a idea what `parse` and `read_blocks` do.

**Comment by h4cc** [comment]
Found a article about "kata driven meetup" that seems usefull für HamburgEx (havent read it yet):

https://medium.com/@mohnishgj/learnings-from-facilitating-a-ruby-meetup-where-people-solved-an-exercism-kata-via-pair-programming-39c78113b86b

**Comment by KlausTrainer** [comment]
Thanks @h4cc for the feedback :)

My intention is that people actually read the private functions and figure out what they do. That said, I'd like to make them as understandable as possible. Adding a few comments is a good hint, and I'll try to address that!

**Comment by KlausTrainer** [comment]
@h4cc Just pushed an update. Maybe you'd like to take a look ;)

**Comment by h4cc** [comment]
@KlausTrainer Easier to understand now, great! But its still a kata for at least intermediate elixir programmers IMHO.

I like that `defmacrop` :)

**Comment by KlausTrainer** [comment]
@h4cc Yes, I also think that it's target audience is intermediate level Elixir programmers.

One could probably make it more accessible by adding some context, however. Maybe give a short introductory talk about Elixir's [`Stream`](http://elixir-lang.org/docs/stable/elixir/Stream.html) module and the [`Enumerable`](http://elixir-lang.org/docs/stable/elixir/Enumerable.html) protocol. Alternatively, one could tell people to study the documentation by themselves, while being available for answering questions with regard to that. [This excellent article about Elixir's continuable enumerators](http://elixir-lang.org/blog/2013/12/11/elixir-s-new-continuable-enumerators/) would also make a great introduction in addition, I guess.

**Comment by KlausTrainer** [comment]
Just added a corresponding [Helpful Resources](https://github.com/hh-ex/kata-dtgcaa#helpful-resources) section to the [dtgcaa kata](/hh-ex/kata-dtgcaa) README.

**Comment by KlausTrainer** [comment]
Also added a [Hints](https://github.com/hh-ex/kata-dtgcaa#hints) section to the dtgcaa kata README, which should make it a bit easier as well.

**Comment by OleMchls** [comment]
@KlausTrainer sorry for the suuuuuuuper slow response, I just added the kata to the list above, and I'm looking forward solving it in one of the next meet ups :) And thanks for the kata, also huge thanks for @h4cc for the great feedback

**Comment by KlausTrainer** [comment]
@nesQuick No worries. Cool! Looking forward to it, too :)

**Comment by h4cc** [comment]
Added a short Kata on binary pattern matching: https://github.com/hh-ex/kata-binary

Solution is in first commit.

**Comment by OleMchls** [comment]
I'd rather move the solution into a separate PR, so people are not
confronted with them when they "just" pull the repo. f you'd add travis CI
for testing and especially PR testing, we can see if the tests pass for the
solutions presented using the github PR UI :)

**Comment by h4cc** [comment]
Idea for a simple Kata: http://www.99-bottles-of-beer.net/

**Comment by benjaminweb** [comment]
https://github.com/hh-ex/planning/pull/90/commits/17c6bd51e75bc4d4c0c3ed1f82c6a2bb1517a3c8
